spec/usecases/reactions/update_materials_spec.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe Usecases::Reactions::UpdateMaterials do
+  let(:user) { create(:user) }
+  let(:collection) { Collection.create!(label: 'Collection', user: user) }
+  let(:reaction) { create(:reaction, name: 'Reaction', collections: [collection]) }
+  let(:root_container) { create(:root_container) }
+  let(:sample) { create(:sample, name: 'Sample', container: FactoryBot.create(:container)) }
+  let(:molfile) { File.read(Rails.root + 'spec/fixtures/test_2.mol') }
+  let(:starting_materials) do
+    {
+      'starting_materials' => [
+        {
+          # hits existing_sample branch
+          'id' => sample.id,
+          'name' => 'starting_material',
+          'target_amount_unit' => 'mg',
+          'target_amount_value' => 75.09596,
+          'equivalent' => 1,
+          'reference' => false,
+          'is_new' => false,
+          'molfile' => molfile,
+          'container' => root_container
+        }
+      ]
+    }
+  end
+  let(:reactants) do
+    {
+      'reactants' => [
+        # hits subsample branch
+        'target_amount_unit' => 'mg',
+        'target_amount_value' => 86.09596,
+        'equivalent' => 2,
+        'reference' => false,
+        'is_new' => true,
+        'molfile' => molfile,
+        'container' => root_container,
+        'parent_id' => sample.id # gets named after parent, hence no name specified
+      ]
+    }
+  end
+  let(:products) do
+    {
+      'products' => [
+        {
+          # hits new_sample branch
+          'name' => 'product',
+          'target_amount_unit' => 'mg',
+          'target_amount_value' => 99.08304,
+          'equivalent' => 5.5,
+          'reference' => false,
+          'is_new' => true,
+          'molfile' => molfile,
+          'container' => root_container
+        }
+      ]
+    }
+  end
+  let(:materials) do
+    {
+      'solvents' => [
+        # hits new_sample branch
+        'name' => 'solvent',
+        'target_amount_unit' => 'mg',
+        'target_amount_value' => 76.09596,
+        'equivalent' => 1,
+        'reference' => true,
+        'is_new' => true,
+        'molfile' => molfile,
+        'container' => root_container
+      ]
+    }.merge(starting_materials, reactants, products)
+  end
+
+  it 'updates existing sample' do
```

Comment:
Please insert a middle layer between the `describe Usecases::Reactions::UpdateMaterials` and the `it` blocks.

Usually the recommended layout of an Rspec file for a class looks like this:
```
describe Usecases::Reactions::UpdateMaterials do
   describe '#execute!' do
      context 'when SOME_CONDITION' do
         it 'does something' do
             ...
         end
     end
  end
end
```

With this structure you can navigate your file easier and some Editor/IDE tooling may even allow you to jump from method definition to the corresponding test and vice versa.